src/totp.rs

```diff
@@ -15,131 +14,162 @@ use base32::Alphabet;
+    /// [`TOTP::from_base32`] constructors.
     secret: Vec<u8>,
 
     /// The digest to use in the HMAC process.
     ///
-    /// Unless an initializer ending in 'with_digest' is used, this value
-    /// defaults to [`MacDigest::SHA1`]
+    /// This value defaults to [`MacDigest::SHA1`] if not specified in a constructor.
     mac_digest: MacDigest,
-}
 
-/// All initializer implementations for the [`TOTP`] struct
-impl TOTP {
-    /// Creates a new TOTP instance with a byte-array representation
-    /// of the secret
+    /// The number of digits of the code generated.
     ///
-    /// Defaults to using [`MacDigest::SHA1`] as the digest for HMAC
-    /// operations.
-    pub fn new(secret: &[u8]) -> Self {
-        TOTP::new_with_digest(secret, MacDigest::SHA1)
-    }
+    /// This value defaults to 6 if not specified in a constructor.
+    digits: u32,
 
-    /// Creates a new TOTP instance with a byte-array representation
-    /// of the secret and a specific digest for HMAC operations
+    /// The period in seconds between two different generated code.
     ///
-    /// Allows for non-SHA1 algorithms to be used with TOTP generation
-    pub fn new_with_digest(secret: &[u8], mac_digest: MacDigest) -> Self {
+    /// This value defaults to 30 if not specified in a constructor.
+    period: u64,
+}
+
+// All initializer implementations for the [`TOTP`] struct
+impl TOTP {
+    /// Generates a new TOTP instance from a byte array representation of the secret,
+    /// a digest algorithm, a number of digits and a period in seconds.
+    pub fn new(secret: &[u8], mac_digest: MacDigest, digits: u32, period: u64) -> Self {
         TOTP {
             secret: secret.to_vec(),
             mac_digest,
+            digits,
+            period,
         }
     }
 
-    /// Creates a new TOTP instance from a utf8-encoded string secret
+    /// Generates a new TOTP instance from an utf8 representation of the secret,
+    /// a digest algorithm, a number of digits and a period in seconds.
+    pub fn new_from_utf8(secret: &str, mac_digest: MacDigest, digits: u32, period: u64) -> Self {
+        TOTP::new(secret.as_bytes(), mac_digest, digits, period)
+    }
+
+    /// Generates a new TOTP instance from a base32 representation of the secret,
+    /// a digest algorithm, a number of digits and a period in seconds.
+    ///
+    /// # Panics
+    /// This method panics if the provided string is not correctly base32 encoded.
+    pub fn new_from_base32(secret: &str, mac_digest: MacDigest, digits: u32, period: u64) -> Self {
+        let decoded = base32_decode(secret).expect("Failed to decode base32 string");
+        TOTP::new(&decoded, mac_digest, digits, period)
+    }
+
+    /// Creates a new TOTP instance with a byte-array representation of the secret.
+    ///
+    /// Defaults to using [`MacDigest::SHA1`] as the digest for HMAC operations,
+    /// 6 digits and a 30 seconds period.
+    pub fn from_secret(secret: &[u8]) -> Self {
```

Comment:
Same thing for these constructors, could we do `default_from_secret` since it's just defaulting to the said values if they arent given